### PR TITLE
Expose join session cached info on request for trello component

### PR DESCRIPTION
**Pull request by jatgarg**
## Description

Expose an api in odsp driver to take driver factory and resolved url and expose the current join session info that is stored in non persistent cache for that document. This was a request from trello component folks, so that they can know about that push session.

### CODE REVIEW

File: packages/drivers/odsp-driver/src/exposeJoinSessionInfo.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * Api which returns the current join session response stored in non persistent cache, if present
+ * @param factory - odsp driver factory
+ * @param resolvedUrl - resolved url for container
+ * @returns - Current join session response stored in cache. Undefined if not present.
+ */
+export async function getJoinSessionInfo(
+	factory: IDocumentServiceFactory,
+	resolvedUrl: IResolvedUrl,
+): Promise<ISocketStorageDiscovery | undefined> {
+	assert(factory instanceof OdspDocumentServiceFactoryCore, "factory type is not recognized");
```
Comment (anthony-murphy) [inline packages/drivers/odsp-driver/src/exposeJoinSessionInfo.ts:22]: this may not work, as instanceof is very particular. even the same type, but from different package version will not satisfy this.

Comment (alexvy86) [inline packages/drivers/odsp-driver/src/exposeJoinSessionInfo.ts:22]: Would throwing a `UsageError` be better here? The factory being something that the consumer passes to a FF API, it feels like a more ergonomic API if we tell the consumer what they did wrong. Based on the scenario we have right now, the component might not be able to do anything to fix the issue at that point, but in a scenario where the host/loader was the one actually calling this instead of making it available to the component, a `UsageError` definitely feels more correct IMO.

Also slightly concerned about the usual `instanceof` caveats. Not sure if the factory instance that the host has could be passed across iframe/window contexts by the time it gets to the component who ends up triggering this call. @vladsud do you know?

File: packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts [hunk 2]
```diff
@@ -56,6 +56,10 @@ export class OdspDocumentServiceFactoryCore implements IDocumentServiceFactory {
 		return this.nonPersistentCache.snapshotPrefetchResultCache;
 	}
 
+	public get joinSessionCache() {
```
Comment (alexvy86) [inline packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts:59]: If we have to expose this publicly on the factory anyway, would it make more sense to just make this the API that returns the `ISocketStorageDiscovery` or undefined and do away with the free function? I know we discussed that the function feels like a weird API for the factory, but to me this feels just as weird, so maybe avoid the extra indirection?

File: packages/drivers/odsp-driver/src/exposeJoinSessionInfo.ts (hunk 1 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/exposeJoinSessionInfo.ts:22]: Since this is inside odsp-driver package, we can maybe just take the argument as type, OdspDocumentServiceFactory.

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/exposeJoinSessionInfo.ts]: Removed instanceof and replaced the type with OdspDriverServiceFactory.

File: packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts (hunk 2 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts:59]: Exposing cache is better than putting this api on factory as its usage is very specific. There could be more such things and it won't make sense to expose such specific api on factory while the cache is something that belongs to factory and could be used for other purpose too. So exposing cache makes much more sense to me than putting this api on factory surface.

File: packages/drivers/odsp-driver/src/contractsPublic.ts [hunk 3]
```diff
@@ -5,6 +5,40 @@
...
+	 */
+	deltaStreamSocketUrl: string;
+
+	/**
+	 * The access token for PushChannel. Optionally returned, depending on implementation.
+	 * OneDrive for Consumer implementation returns it and OneDrive for Business implementation
+	 * does not return it and instead expects token to be returned via `getWebsocketToken` callback
+	 * passed as a parameter to `OdspDocumentService.create()` factory.
+	 */
+	socketToken?: string;
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/contractsPublic.ts:34]: Given that we touch this area, could you please follow up with Marcelo if this thing should be deleted (or maybe vice versa - become non-optional)? I do not remember the history well, but I think it was temp thing, maybe it's time to clean it (if we can)

File: packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts (hunk 2 above)
Comment (vladsud) [reply packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts:59]: Just to be clear, while this is public method, it's usage (ideally) should be constrained to driver. I believe we have \@internal mark-up, if so, I'd use it on this API, to discourage any usage by hosts (who have access to a concrete class).
That's a bit beyond this PR, but we should consider not exposing these classes at all, instead expose flat factory function (that returns IDocumentServiceFactory)

Comment (vladsud) [reply packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts:59]: That said, I think I agree with Alejandro.

File: packages/drivers/odsp-driver/src/exposeJoinSessionInfo.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * Licensed under the MIT License.
+ */
+
+import { IResolvedUrl } from "@fluidframework/driver-definitions";
+import { ISocketStorageDiscovery } from "./contractsPublic";
+import { getJoinSessionCacheKey, getOdspResolvedUrl } from "./odspUtils";
+import { OdspDocumentServiceFactory } from "./odspDocumentServiceFactory";
+
+/**
+ * Api which returns the current join session response stored in non persistent cache, if present
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/exposeJoinSessionInfo.ts:12]: While this description is good, I think it's hard to parse by users (who have no knowledge of the internals of the driver). I'd add here a bit more using terminology that is more targeted to users of this API.
I.e., this API would return info about relay service session only if this factory established (or attempted to establish) connection very recently. Otherwise, it will return undefined.

**Review by vladsud — APPROVED** [review]

File: packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts (hunk 2 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts:59]: If you both agree, then I can put this on the factory

File: packages/drivers/odsp-driver/src/contractsPublic.ts (hunk 3 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/contractsPublic.ts:34]: I will follow up on this separately. Seems like when getWebsocketToken callback is not supplied by the app, then we look for the socket token from join session. I am not sure whether they always supply this or not.
